### ISSUE TITLE
Support CUDA autocast for `torch.linalg.vecdot()`

### ISSUE THREAD

**Issue by gau-nernst** [pytorch/pytorch#108127]
### 🚀 The feature, motivation and pitch

Currently `torch.linalg.vecdot()` does not work with CUDA autocast. Documentation also does not mention that vecdot is supported: https://pytorch.org/docs/stable/amp.html#cuda-ops-that-can-autocast-to-float16

```python
import torch

x = torch.randn(2, 100, 256, device="cuda", dtype=torch.float16)
y = torch.randn(2, 100, 256, device="cuda")

with torch.autocast("cuda", torch.float16):
    out = torch.linalg.vecdot(x, y)
    # out = (x.unsqueeze(-2) @ y.unsqueeze(-1)).squeeze(-1)  # this works
    # out = (x * y).sum(-1)  # this works, but promotes x to float32 and calculates in float32
    # out = torch.einsum("...i,...i->...", x, y)  # this works too

print(out.dtype)
```

### Alternatives

Workarounds (included in code snippet above):
- Adjust dimensions and use matmul: looks pretty ugly, hard to understand, and may be prone to errors
- Point-wise multiply and sum: easy to understand, but materialize an intermediate tensor. Also, calculation is performed in float32
- `torch.einsum()`: works well. I haven't benchmarked the speed to see if it's slower or faster. Although it's also easy to understand, using `vecdot()` is still slightly clearer.

cc @ptrblck @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano @mcarilli @leslie-fang-intel @jgong5

**Comment by vadimkantorov** [comment]
Btw, is `torch.linalg.vecdot(x, x)` faster/memory-economical than doing `x.pow(2).sum(dim = -1)`? it would be nice if it is!

**Comment by lezcano** [comment]
It doesn't do anything interesting: 
https://github.com/pytorch/pytorch/blob/fe2cda64dc59504d2d121e2d0a92f0eed9aed803/aten/src/ATen/native/BatchLinearAlgebra.cpp#L3796-L3807
so the OP is wrong in saying that 
>Point-wise multiply and sum: easy to understand, but materialize an intermediate

as `vecdot` also materialises an intermediate.

Now, if you torch-compile `(x * x).sum(dim=-1)` you'll certainly see a speed-up.

**Comment by gau-nernst** [comment]
Interesting... I didn't know linalg.vecdot() also performs multiply and sum. I always have the wrong impression that it uses an optimized way to calculate dot product.
Are there any plans to optimize linalg.vecdot() further? E.g. not realizing an intermediate tensor.

**Comment by lezcano** [comment]
You can do that by `torch.compile`-ing `(x * y).sum(-1)`, for example.
I put up two PRs at https://github.com/pytorch/pytorch/pull/108194, after which you'll be able to simply `torch.compile` `linalg.vecdot` and get an efficient code that does not have any intermediary tensor. You try this with the 1D version already by running
```python
import torch
import numpy as np


@torch.compile()
def fn(x, y):
    return torch.dot(x, y)


x = torch.randn(512, device="cuda")
y = torch.randn(512, device="cuda")
fn(x,y)
```
with `TORCH_LOGS=outputcode`.

We don't have any plans to implement this optimisation in eager, really.

**Comment by vadimkantorov** [comment]
@lezcano It would be great if PyTorch core simply dispatched vecdot(x, x) or vecdot(x, y) to the compiled implementation.

Currently I found that compiler outputs are not very reliable because Dynamo options are global: https://github.com/pytorch/pytorch/issues/106614 :( I think for reliable inclusion of these things in libraries (and especially in core libraries), there needs to be a way of local compiler options (including dynamic shape specification)

**Comment by gau-nernst** [comment]
I agree with @vadimkantorov, that it will be great if PyTorch core uses an optimized version of `vecdot()` out-of-the-box (in eager mode). I guess `vecdot()` is more for complex numbers, because it seems like there are no benefits in using `vecdot()` for real numbers (over alternatives like multiply-and-sum, matmul, or einsum)?

**Comment by lezcano** [comment]
I agree that that'd be better, but I think we are a bit far from having anything like that in the short term. I would resort to `torch.compile` if this operation is a bottleneck in your program.

**Comment by gau-nernst** [comment]
I see. Thank you for the clarification and the quick fix to the original issue!

**Comment by vadimkantorov** [comment]
> I agree that that'd be better, but I think we are a bit far from having anything like that in the short term.

Would inductor-fused optimizers maybe find the way to the core first?

**Comment by gau-nernst** [comment]
@vadimkantorov The original issue has been resolved. I think if you have any other questions, you should open a new issue or discussion. Thank you.

On a related note, I do believe latest PyTorch has fused optimizer implementation.
